grid/chat/messages/maps/slack.suma

```diff
@@ -0,0 +1,243 @@
+				statusCode = statusCode,
+				error = body,
+				headers = headers,
+			)
+			
+			map error error
+		}
+	}
+}
+
+operation MapMessages {
+	set if (args.messages && args.messages.length > 0) {
+		messages = args.messages.filter(message => message.type === 'message')
+	}
+
+	listOfMessages = []
+	
+	// workaround for map interpreter variables merging issue
+	listOfMessages = call foreach(index of Array(messages.length).fill().map((_, i) => i)) MapMessage(
+		index = index,
+		messages = messages,
+		dest = args.destination,
+	) if (messages.length > 0)
+
+	return listOfMessages
+}
+
+operation MapMessage {
+	index = args.index
+	message = args.messages[index]
+
+	messageResult = {
+		id: message.ts,
+		text: message.text,
+		createdAt: parseFloat(message.ts),
+	}
+
+	set if(message.user || message.bot_id) {
+		messageResult.author = {
+			id: message.user || message.bot_id,
+			username: message.username
+		}
+	}
+
+	set if (message.thread_ts) {
+		messageResult.hasThread = true
+		messageResult.threadId = message.thread_ts;
+	}
+
+	set if (message.edited) {
+		messageResult.updatedAt = parseFloat(message.edited.ts);
+	}
+
+	set if (message.reactions) {
+		messageResult.reactions = message.reactions.map(reaction => ({
+			emoji: reaction.name,
+			count: reaction.count,
+			users: reaction.users
+		}))
+	}
+
+	set if (message.files && message.files.length > 0) {
+		messageResult.attachments = message.files.map(file => ({
+			id: file.id,
+			createdAt: file.created,
+			fileName: file.name,
+			mediaType: file.mimetype,
+			url: file.permalink_public,
+			preview: file.preview,
+		}))
+	}
+
+	messageResult.url = call ComposeURL(ts = message.ts, dest = args.dest)
+
+	return messageResult
+}
+
+"""
+Operation MapRateLimitDetails
+Returns object mapping rate limit details specified in received headers
+"""
+operation MapRateLimitDetails {
+	headers = args.headers
+	details = {}
+
+	totalRequests = parseInt(headers['X-Rate-Limit-Limit'])
+	remainingRequests = parseInt(headers['X-Rate-Limit-Remaining'])
+	resetAfter = parseInt(headers['X-Rate-Limit-Reset'])
+	retryAfter = parseInt(headers['Retry-After'])
+
+	set if (!isNaN(totalRequests)) {
+		details.totalRequests = totalRequests
+	}
+
+	set if (!isNaN(remainingRequests)) {
+		details.remainingRequests = remainingRequests
+	}
+
+	set if (!isNaN(resetAfter)) {
+		details.resetAfter = resetAfter
+		details.resetTimestamp = Date.now() + (resetAfter * 1000)
+	}
+
+	set if (!isNaN(retryAfter)) {
+		details.retryAfter = retryAfter
+	}
+
+	set if (details.remainingRequests && details.totalRequests) {
+		details.remainingRequestsPercentage = (details.remainingRequests / details.totalRequests) * 100
+	}
+
+	set if (Object.keys(details).length === 0) {
+		details = undefined
+	}
+
+	return details
+}
+
+operation ComposeURL {
+	timestamp = args.ts
+	destination = args.dest
+
+	return `https://slack.com/api/conversations.history?channel=${destination}&latest=${timestamp}&inclusive=true&limit=1`
```

Comment:
Huh, if this is publicly exposed URL, I don't think it should point to an API, but maybe I misunderstand how this is used?

Reply:
Yeah, this points to the API 😕